docs/_posts/2015-02-24-streamlining-react-elements.md

```diff
@@ -0,0 +1,230 @@
+
+```js
+var foo = <input className="foo" />;
+class Component {
+  render() {
+    return bar ? <input className="bar" /> : foo;
+  }
+}
+```
+
+These two inputs have different owners, therefore React will not keep its state when the conditional switches. There is nothing in the code to indicate that. Similarly, if you use `React.addons.cloneWithProps`, the owner changes.
+
+### Problem: Timing Matters
+
+The owner is tracked by the currently executing stack. This means that the semantics of a ReactElement varies depending on when it is executed. Take this example:
+
+```js
+class A {
+  render() {
+    return <B renderer={text => <span>{text}</span>} />;
+  }
+}
+class B {
+  render() {
+    return this.props.renderer(‘foo’);
+  }
+}
+```
+
+The owner of the `span` is actually `B`, not `A` because of the timing of the callback. This is all adds complexity and suffers from similar problems as mutation.
+
+### Problem: It Couples JSX to React
+
+Have you wondered why JSX depends on React? Couldn’t the transpiler have that built-in to its runtime? The reason you need to have `React.createElement` in scope is because we depend on internal state of React to capture the current "owner". Without this, you wouldn’t need to have React in scope.
+
+### Solution: Make Context Parent-Based Instead of Owner-Based
+
+The first thing we’re doing is warning you if you’re using the "owner" feature in a way that relies on it propagating through owners. Instead, we’re planning on propagating it through parents to its children. In almost all cases, this shouldn’t matter. In fact, parent-based contexts is simply a superset.
+
+### Solution: Remove the Semantic Implications of Owner
+
+It turns out that the cases where owners as part of state-semantics. As a precaution, we’ll warn you if it turns out that the owner is important to determine state. In almost every case this shouldn’t matter. Unless you’re doing some weird optimizations, you shouldn’t see this warning.
+
+### Pending: Change the refs Semantics
+
+Refs are still based on "owner". We haven’t fully solved this special case just yet.
+
+In 0.13 we introduced a new callback-refs API that doesn’t suffer from these problems but we’ll keep on a nice declarative alternative to the current semantics for refs. As always, we won’t deprecate something until we’re sure that you’ll have a nice upgrade path.
+
+## Keyed Objects as Maps
+
+In React 0.12 you could use keyed objects to provide an external key to an element or a set. This pattern isn’t actually widely known, nor documented outside of Facebook. It shouldn’t be an issue for most of you.
+
+```js
+<div>{{ a: <span />, b: <span /> }}</div>
+```
+
+### Problem: Relies on Enumeration Order
+
+The problem with this pattern is that it relies on enumeration order of objects. This is technically unspecified, even though implementations now agree to use insertion order. Except for the special case when numeric keys are used.
```

Comment:
Perhaps can say that we've had a warning for this for a while so you shouldn't be using numeric keys anyway.

There's also the fun old IE case with `delete` and reinsertion of an old key going to the old position in enumeration, not the end.